### PR TITLE
issue #126 - Clean up main()

### PR DESCRIPTION
**Pull request by stonecoldhughes**
Fixes issue #126 , which was outdated. Cleaned up main by moving combine_dimensions blocks into a function.

### CODE REVIEW

**Comment by grahamlopez** [comment]
Please make the title of pull requests say what they do, and if they are fixing an issue, include that in the description (which should also expand on the PR title).

**Review by bmcdanie — COMMENTED** [review]
+67/-18 is not a great net for a DRY'ing out PR. hmm...

may need to rethink changing this, or have the pass in the function arrays (`source.source_funcs[i]`, for instance) so that we can extract more of the shared code. I'm in CASL today if you want to discuss.

**Comment by bmcdanie** [comment]
actually, it looks like the PDE holds these functions in vectors already - is it redundant to repackage them? they can probably be passed straight in.

**Comment by bmcdanie** [comment]
example:

```  
  121   // -- generate source vectors.
  122   // these will be scaled later according to the simulation time applied
  123   // with their own time-scaling functions
  124   node_out() << "  generating: source vectors..." << '\n';
  125   std::vector<fk::vector<prec>> const initial_sources = [&pde, &table, &subgrid,
  126                                                          degree]() {
  127     std::vector<fk::vector<prec>> initial_sources;
  128     for (source<prec> const &source : pde->sources)
  129     {
  130       initial_sources.push_back(transform_and_combine_dimensions(
  131           *pde, source.source_funcs, table, subgrid.row_start, subgrid.row_stop,
  132           degree));
  133     }
  134     return initial_sources;
  135   }();
```

File: src/transformations.hpp [hunk 1]
```diff
@@ -31,6 +31,24 @@ fk::vector<P>
 combine_dimensions(int const, element_table const &,
                    std::vector<fk::vector<P>> const &, P const = 1.0);
 
+template<typename P>
+fk::vector<P>
+transform_and_combine_dimensions(PDE<P> const &pde,
```
Comment (bmcdanie) [inline src/transformations.hpp]: let's accept the functions by const ref and const the int args here

File: src/main.cpp [hunk 2]
```diff
@@ -104,16 +104,18 @@ int main(int argc, char **argv)
 
   // -- generate initial condition vector.
   node_out() << "  generating: initial conditions..." << '\n';
+
   fk::vector<prec> const initial_condition = [&pde, &table, &subgrid,
                                               degree]() {
-    std::vector<fk::vector<prec>> initial_conditions;
+    std::vector<vector_func<prec>> v_functions;
```
Comment (bmcdanie) [inline src/main.cpp:110]: trim this unnecessary step if possible, see my example

File: src/transformations.cpp [hunk 3]
```diff
@@ -9,6 +9,35 @@
...
 #include <numeric>
 
+template<typename P>
+fk::vector<P>
+transform_and_combine_dimensions(PDE<P> const &pde,
+                                 std::vector<vector_func<P>> v_functions,
+                                 element_table const &table, int start,
+                                 int stop, int degree)
+{
+  std::vector<fk::vector<P>> dimension_components;
```
Comment (bmcdanie) [inline src/transformations.cpp]: assert stop and start >= 0, stop >= start, degree  > 0, etc.

Comment (bmcdanie) [reply src/transformations.cpp]: also that the v_functions match num pde dims

**Review by bmcdanie — CHANGES_REQUESTED** [review]
add these few items and let's get a simple test for this in transformations tests since it's a public function.

File: src/transformations.hpp (hunk 1 above)
Comment (stonecoldhughes) [reply src/transformations.hpp]: done

File: src/transformations.cpp (hunk 3 above)
Comment (stonecoldhughes) [reply src/transformations.cpp]: done

File: src/main.cpp (hunk 2 above)
Comment (bmcdanie) [reply src/main.cpp:110]: can't, because no vector of functions directly available here as in the source case. need to consider in pde spec rewrite.

**Review by bmcdanie — APPROVED** [review]
thanks.

**Pull request merged**